Add adaptive profile QIR mid-circuit measurement .ll examples and corresponding Executor tests

OsamaMIT opened the pull request:
This pull request contributes two new QIR example files compliant with the under-development adaptive profile specification, showcasing mid-circuit measurements and classical control flow constructs such as conditionals and loops. These examples follow the QIR specification and expand the test suite with practical adaptive quantum programs.

Additionally, corresponding test cases have been added to test/qiree/Executor.test.cc, integrating these examples into the existing Executor test framework. The test cases integrate the new QIR examples into the existing Executor test framework, using the existing QuantumTestImpl interface.

These additions address issue [#31](https://github.com/ORNL-QCI/qiree/issues/31) by enhancing test coverage and providing canonical QIR adaptive profile examples useful for ongoing development and validation of the QIR runtime.

sethrj commented:
@OsamaMIT Thank you for your contribution! Can you please:
1. Is there a box "Allow edits and access to secrets by maintainers" that you can check? We can't push corrections to your topic branch.
2. The tests don't currently pass. First you need to move the tests to `test/data` and fix the path names. Second, run the tests and you'll see the output is not correct.

OsamaMIT commented:
I replaced the quantum algorithms in an attempt to provide more useful algorithms. Introduced Kitaev Phase Estimation: one qubit phase estimation circuit, 3-qubit bit flip error correction: quantum error correction. I attempted to match the theory for these algorithms in the circuit according to my understanding.

I also have issues with locally running the tests but I will work through them until the review is successful.

OsamaMIT commented:
I used PyQIR but made some manual changes in an attempt to change the circuit logic

wongey commented:
@OsamaMIT  -- the examples are running on my end now!

The bit_flip example doesn't seem to do anything and I always get the result 000. Is this the intention? What is the purpose of this example? 

Regarding phase estimation -- this seems to be an implementation of one round of iterative phase estimation. Is that right? If this is the intention, there may be a H gate on qubit 0 missing (right before measurement). In principle, there would be multiple rounds to get better accuracy (with the angle changing at each round). This may require a qubit reset before the next round. A full example would include the second round.

OsamaMIT commented:
In the bit flip example I don't inject an error (flip a bit) so it always returns the same data measurement only to avoid complicating tests. If preferred I could change it so that it always flips and therefore must always fix the error but still return the same measurement to simplify tests. 

For the phase estimation I'll add the missing H gate along with resetting the qubit to setup two rounds.

wongey commented:
Thank you @OsamaMIT for producing two new QIR-compatible examples that showcase the use of mid-circuit measurement! We appreciate your responsiveness and willingness to engage with us for the Unitary Hack 2025 event. We plan to use your work as a starting point to develop more complicated examples.
